boa_engine/src/environments/runtime.rs

```diff
@@ -852,3 +843,323 @@ impl BindingLocator {
+                    } else if !env.with.get() {
+                        break;
+                    }
+                }
+                Environment::Object(o) => {
+                    let o = o.clone();
+                    let key: JsString = self
+                        .interner()
+                        .resolve_expect(name.sym())
+                        .into_common(false);
+                    if o.has_property(key.clone(), self)? {
+                        if let Some(unscopables) = o.get(JsSymbol::unscopables(), self)?.as_object()
+                        {
+                            if unscopables.get(key.clone(), self)?.to_boolean() {
+                                continue;
+                            }
+                        }
+                        return o.set(key, value, true, self);
+                    }
+                }
+            }
+        }
+
+        let mut bindings = self
+            .realm
+            .environments
+            .stack
+            .get(environment_index)
+            .expect("environment index must be in range")
+            .as_declarative()
+            .expect("msg")
+            .bindings
+            .borrow_mut();
+        let binding = bindings
+            .get_mut(binding_index)
+            .expect("binding index must be in range");
+        if binding.is_none() {
+            Ok(false)
+        } else {
+            *binding = Some(value);
+            Ok(true)
+        }
+    }
+
+    /// Set the value of a binding by it's name.
+    ///
+    /// This only considers function environments that are poisoned.
+    /// All other bindings are set via indices.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the environment or binding index are out of range.
+    pub(crate) fn put_value_if_global_poisoned(
+        &mut self,
+        name: Identifier,
+        value: &JsValue,
+    ) -> JsResult<bool> {
+        for env_index in (0..self.realm.environments.stack.len()).rev() {
+            let env = self
+                .realm
+                .environments
+                .stack
+                .get(env_index)
+                .expect("environment index must be in range");
+
+            match env {
+                Environment::Declarative(env) => {
+                    if env.poisoned.get() {
+                        let compile = env.compile.borrow();
+                        if compile.is_function() {
+                            if let Some(b) = compile.get_binding(name) {
+                                let mut bindings = self
+                                    .realm
+                                    .environments
+                                    .stack
+                                    .get(b.environment_index)
+                                    .expect("environment index must be in range")
+                                    .as_declarative()
+                                    .expect("environment must be declarative")
```

Comment:
```suggestion
                                let mut bindings = self
                                    .declarative_environment_expect(b.environment_index)
```
I would also refactor this into a utility function